### PR TITLE
empty: init at 0.6.21b

### PR DESCRIPTION
**Pull request by djwf**
The program `empty` is not in nixpkgs.

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [X] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 116067` at 6ea7f7c1 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build (new failures):</summary>
  <ul>
    <li><a href="https://gist.github.com/b67681e29e204a3f4877b2ec970d12fa">empty</a></li></li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

which is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/tools/misc/empty/default.nix:18:3:
```
   |
18 |   buildInputs = [ which ];
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/tools/misc/empty/default.nix:24:3:
```
   |
24 |   meta = with lib; {
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/tools/misc/empty/default.nix:22:15:
```
   |
22 |   patches = [ ./allow-prefix-override.patch ./fix-cc-parameter-order.patch ];
   |               ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/tools/misc/empty/default.nix:22:45:
```
   |
22 |   patches = [ ./allow-prefix-override.patch ./fix-cc-parameter-order.patch ];
   |                                             ^
```
</li>
  </ul>
</details>

File: pkgs/tools/misc/empty/allow-prefix-override.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+diff --git a/Makefile b/Makefile
+index 612cc23..a198039 100644
+--- a/Makefile
++++ b/Makefile
+@@ -13,7 +13,7 @@
+ CC =	cc
+ LIBS =	-lutil
+ 
+-PREFIX = /usr/local
++PREFIX ?= /usr/local
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/empty/allow-prefix-override.patch:10]: Please convert this patch to subsituteInPlace.

File: pkgs/tools/misc/empty/default.nix [hunk 2]
```diff
@@ -0,0 +1,49 @@
+{ fetchgit, lib, stdenv, which }:
+
+let
+  baseName = "empty";
+  rev = "6e81f477dba6c2f9086da70b02abec71b64fb16e";
+  baseVersion = "0.6.21b";
+  version = baseVersion + "+1-git-" + (builtins.substring 0 7 rev);
+  name = baseName + "-" + version;
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/empty/default.nix:8]:
```suggestion
```

File: pkgs/tools/misc/empty/default.nix [hunk 3]
```diff
@@ -0,0 +1,49 @@
+{ fetchgit, lib, stdenv, which }:
+
+let
+  baseName = "empty";
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/empty/default.nix:4]:
```suggestion
  pname = "empty";
```

File: pkgs/tools/misc/empty/default.nix [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  rev = "6e81f477dba6c2f9086da70b02abec71b64fb16e";
+  baseVersion = "0.6.21b";
+  version = baseVersion + "+1-git-" + (builtins.substring 0 7 rev);
+  name = baseName + "-" + version;
+in stdenv.mkDerivation {
+  inherit name;
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/empty/default.nix:10]:
```suggestion
  inherit pname version;
```

File: pkgs/tools/misc/empty/default.nix [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+  src = fetchgit {
+    inherit rev;
+    url = "https://git.code.sf.net/p/${name}/code";
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/empty/default.nix:14]:
```suggestion
    url = "https://git.code.sf.net/p/${pname}-${version}/code";
```

File: pkgs/tools/misc/empty/default.nix [hunk 6]
```diff
@@ -0,0 +1,49 @@
+{ fetchgit, lib, stdenv, which }:
+
+let
+  baseName = "empty";
+  rev = "6e81f477dba6c2f9086da70b02abec71b64fb16e";
+  baseVersion = "0.6.21b";
+  version = baseVersion + "+1-git-" + (builtins.substring 0 7 rev);
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/empty/default.nix:7]: Is that an actual release or some commit after a release?

File: pkgs/tools/misc/empty/default.nix [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  makeFlags = [ "PREFIX=$(out)" ];
+
+  # Allow the installation previx to be overridden, and fix the CC parameter
+  # order.
+  patches = [ ./allow-prefix-override.patch ./fix-cc-parameter-order.patch ];
+
+  meta = with lib; {
+    homepage = "https://empty.sourceforge.net";
+    description =
+      "A program to run processes and applications under pseudo-terminal (PTY) sessions and replace TCL/Expect with a simple shell-tool";
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/empty/default.nix:29]:
```suggestion
    description = "A program to run processes and applications under pseudo-terminal (PTY) sessions and replace TCL/Expect with a simple shell-tool";
```

File: pkgs/tools/misc/empty/default.nix [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+      tools because empty:
+
+      - can be easily invoked directly from shell prompt or script
+      - does not use TCL, Perl, PHP, Python or anything else as an underlying language
+      - is written entirely in C
+      - has small and simple source code
+      - can easily be ported to almost all UNIX-like systems
+    '';
+    maintainers = [ maintainers.djwf ];
+    priority = 10;
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/empty/default.nix:47]: Can you add a comment why we want this?

File: pkgs/tools/misc/empty/fix-cc-parameter-order.patch [hunk 9]
```diff
@@ -0,0 +1,13 @@
+diff --git a/Makefile b/Makefile
+index a198039..c87b5d5 100644
+--- a/Makefile
++++ b/Makefile
+@@ -18,7 +18,7 @@ BINDIR = $(PREFIX)/bin
+ MANDIR = $(PREFIX)/man/man1
+ 
+ all:
+-	${CC} ${CFLAGS} -Wall ${LIBS} -o empty empty.c
++	${CC} ${CFLAGS} -Wall -o empty empty.c ${LIBS}
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/empty/fix-cc-parameter-order.patch:10]: This patch can also be easily replaced by substituteInPlace.

File: pkgs/tools/misc/empty/default.nix (hunk 6 above)
Comment (djwf) [reply pkgs/tools/misc/empty/default.nix:7]: It's a commit after a release: the release is 0.6.21b, the +1 is one commit after the release, the git-rev substring is the commit hash of that commit. Should I label it in a different way?

File: pkgs/tools/misc/empty/fix-cc-parameter-order.patch (hunk 9 above)
Comment (djwf) [reply pkgs/tools/misc/empty/fix-cc-parameter-order.patch]: Done.

File: pkgs/tools/misc/empty/default.nix (hunk 2 above)
Comment (djwf) [reply pkgs/tools/misc/empty/default.nix]: Done.

File: pkgs/tools/misc/empty/default.nix (hunk 4 above)
Comment (djwf) [reply pkgs/tools/misc/empty/default.nix]: Done.

File: pkgs/tools/misc/empty/default.nix (hunk 5 above)
Comment (djwf) [reply pkgs/tools/misc/empty/default.nix]: This won't work - the repo is based on the name of the project (empty), and there's no version in the path.

**Comment by djwf** [comment]
Is there anything else I need to do to get this merged? Or is it just a matter of time?

File: pkgs/tools/misc/empty/default.nix (hunk 6 above)
Comment (SuperSandro2000) [reply pkgs/tools/misc/empty/default.nix:7]: We generally use upstream release numbers minus the v or the unstable prefix with the date of the commit.

File: pkgs/tools/misc/empty/default.nix [hunk 10]
```diff
@@ -0,0 +1,50 @@
+{ fetchgit, lib, stdenv, which }:
+
+let
+  pname = "empty";
+  rev = "6e81f477dba6c2f9086da70b02abec71b64fb16e";
+  baseVersion = "0.6.21b";
+  version = baseVersion + "+1-git-" + (builtins.substring 0 7 rev);
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/empty/default.nix:7]:
```suggestion
  version = "unstable-XXXX-XX-XX";
```
where XXXX-XX-XX the date of the commit.

File: pkgs/tools/misc/empty/default.nix [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+  # Allow installation prefix to be overridden, and fix the CC parameter order.
+  prePatch = ''
+    substituteInPlace Makefile \
+      --replace 'PREFIX = ' 'PREFIX ?= ' \
+      --replace '\$\{LIBS\} -o empty empty.c' '-o empty empty.c \$\{LIBS\}'
+  '';
+
+  meta = with lib; {
+    homepage = "https://empty.sourceforge.net";
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/empty/default.nix:29]: Page refuses connection for me.

File: pkgs/tools/misc/empty/default.nix (hunk 8 above)
Comment (SuperSandro2000) [reply pkgs/tools/misc/empty/default.nix:48]: So this program also contains the expect binary which conflicts with expect?

File: pkgs/tools/misc/empty/default.nix [hunk 12]
```diff
@@ -0,0 +1,50 @@
+{ fetchgit, lib, stdenv, which }:
+
+let
+  pname = "empty";
+  rev = "6e81f477dba6c2f9086da70b02abec71b64fb16e";
+  baseVersion = "0.6.21b";
+  version = baseVersion + "+1-git-" + (builtins.substring 0 7 rev);
+in stdenv.mkDerivation {
+  inherit pname version;
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/empty/default.nix:9]:
```suggestion
stdenv.mkDerivation rec {
  pname = "empty";
  version = "unstable-XXXX-XX-XX";
```
where XXXX-XX-XX the date of the commit.

File: pkgs/tools/misc/empty/default.nix (hunk 6 above)
Comment (djwf) [reply pkgs/tools/misc/empty/default.nix:7]: Ok: as the single additional commit is not critical (especially in a packaged context), I'll package the version.

File: pkgs/tools/misc/empty/default.nix (hunk 8 above)
Comment (djwf) [reply pkgs/tools/misc/empty/default.nix:48]: Nope: it contains the binary "empty" that performs tasks similar to expect, but using a different interface. There's a really good example on the home page (http://empty.sourceforge.net) that shows examples using empty, TCL/Expect, Perl/expect module, and Python/Pexpect.

File: pkgs/tools/misc/empty/default.nix (hunk 11 above)
Comment (djwf) [reply pkgs/tools/misc/empty/default.nix:29]: Ah - the problem is the "s", fixing.

File: pkgs/tools/misc/empty/default.nix (hunk 12 above)
Comment (djwf) [reply pkgs/tools/misc/empty/default.nix]: Packaged the latest release instead of the latest commit.

**Comment by djwf** [comment]
Alright - I believe I've addressed all of the issues raised above - please let me know if there's anything else I need to adjust. Thanks!

**Comment by zseri** [comment]
you should probably squash the latest commits into one.

File: pkgs/tools/misc/empty/default.nix [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  };
+
+  nativeBuildInputs = [ which ];
+
+  makeFlags = [ "PREFIX=$(out)" ];
+
+  # Allow installation prefix to be overridden, and fix the CC parameter order.
+  prePatch = ''
+    substituteInPlace Makefile \
+      --replace 'PREFIX = ' 'PREFIX ?= ' \
```
Comment (zseri) [inline pkgs/tools/misc/empty/default.nix:22]: Have you tested the package without this line? I suspect the make variant which nixpkgs uses doesn't require it.

Comment (djwf) [reply pkgs/tools/misc/empty/default.nix:22]: Thanks: that's good to know!

**Comment by djwf** [comment]
> you should probably squash the latest commits into one.

Got it: I'll squash all the commits into the standard "empty: init at 0.6.21b"

**Comment by djwf** [comment]
Ok - squashed and pushed. Let me know if anything else needs to be done. Thanks for the guidance!

File: pkgs/tools/misc/empty/default.nix (hunk 8 above)
Comment (SuperSandro2000) [reply pkgs/tools/misc/empty/default.nix:45]:
```suggestion
```

Then I don't think we need this.

File: pkgs/tools/misc/empty/default.nix [hunk 14]
```diff
@@ -0,0 +1,47 @@
+{ fetchzip, lib, stdenv, which }:
+
+let
+  pname = "empty";
+  version = "0.6.21b";
+in stdenv.mkDerivation {
+  inherit pname version;
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/empty/default.nix:7]:
```suggestion
stdenv.mkDerivation rec {
  pname = "empty";
  version = "0.6.21b";
```

Comment (djwf) [reply pkgs/tools/misc/empty/default.nix]: Done and pushed.

**Comment by djwf** [comment]
Is there anything else that I need to do?

File: pkgs/tools/misc/empty/default.nix [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+  };
+
+  nativeBuildInputs = [ which ];
+
+  makeFlags = [ "PREFIX=$(out)" ];
+
+  # Allow installation prefix to be overridden, and fix the CC parameter order.
+  prePatch = ''
+    substituteInPlace Makefile \
+      --replace '\$\{LIBS\} -o empty empty.c' '-o empty empty.c \$\{LIBS\}'
```
Comment (zseri) [inline pkgs/tools/misc/empty/default.nix:20]: > empty> substituteStream(): WARNING: pattern '\$\{LIBS\} -o empty empty.c' doesn't match anything in file 'Makefile'
```suggestion
      --replace '${LIBS} -o empty empty.c' '-o empty empty.c ${LIBS}'
```

**Review by fogti — APPROVED** [review]

File: pkgs/tools/misc/empty/default.nix (hunk 15 above)
Comment (djwf) [reply pkgs/tools/misc/empty/default.nix]: Gotcha - thanks for the pull request - it's been merged!

**Comment by djwf** [comment]
Is there anything else that I need to do?

File: maintainers/maintainer-list.nix [hunk 16]
```diff
@@ -2375,6 +2375,12 @@
     githubId = 10913120;
     name = "Dje4321";
   };
+  djwf = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:2378]: Please put this into a separate commit with the message ``maintainers: add NAME`` and sort it alphabetically.

File: pkgs/tools/misc/empty/0.6-Makefile.patch [hunk 17]
```diff
@@ -0,0 +1,13 @@
+diff --git a/Makefile b/Makefile
+index 1fe4c41..2c69558 100644
+--- a/Makefile
++++ b/Makefile
+@@ -16,7 +16,7 @@ LIBS =	-lutil
+ PREFIX = /usr/local
+ 
+ all:
+-	${CC} ${CFLAGS} -Wall ${LIBS} -o empty empty.c
++	${CC} ${CFLAGS} -Wall -o empty empty.c ${LIBS}
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/empty/0.6-Makefile.patch:10]: This could be substituteInPlace but not that important.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Also please remove the merge commit from the history and squash all package related commits together.

File: pkgs/tools/misc/empty/0.6-Makefile.patch (hunk 17 above)
Comment (zseri) [reply pkgs/tools/misc/empty/0.6-Makefile.patch:10]: it could be, but it is suprisingly hard to get right.

File: maintainers/maintainer-list.nix (hunk 16 above)
Comment (djwf) [reply maintainers/maintainer-list.nix:2378]: Both are done.

File: pkgs/tools/misc/empty/0.6-Makefile.patch (hunk 17 above)
Comment (djwf) [reply pkgs/tools/misc/empty/0.6-Makefile.patch:10]: Leaving it be.

**Comment by SuperSandro2000** [comment]
Sorry that this took so long.
